store/v2/multi/snapshot.go

```diff
@@ -0,0 +1,259 @@
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, fmt.Sprintf("error while get the version at height %d", height))
+	}
+
+	// Spawn goroutine to generate snapshot chunks and pass their io.ReadClosers through a channel
+	ch := make(chan io.ReadCloser)
+	go func() {
+		// Set up a stream pipeline to serialize snapshot nodes:
+		// Export KV Item -> delimited Protobuf -> zlib -> buffer -> chunkWriter -> chan io.ReadCloser
+		chunkWriter := snapshots.NewChunkWriter(ch, snapshotChunkSize)
+		defer chunkWriter.Close()
+		bufWriter := bufio.NewWriterSize(chunkWriter, snapshotBufferSize)
+		defer func() {
+			if err := bufWriter.Flush(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		// zlib compression levels:  https://www.euccas.me/zlib/#zlib_compression_levels
+		// zlib default compression level 6
+		// level 0 :  no compression and fastest
+		// ....
+		// level 7 : average compression and average speed
+		// ...
+		// level 9 : highest compression and speed is slower
+		zWriter, err := zlib.NewWriterLevel(bufWriter, 7)
+		if err != nil {
+			chunkWriter.CloseWithError(sdkerrors.Wrap(err, "zlib failure"))
+			return
+		}
+		defer func() {
+			if err := zWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		protoWriter := protoio.NewDelimitedWriter(zWriter)
+		defer func() {
+			if err := protoWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+
+		var sKeys []string
+		// sending the snapshot store schema
+		for sKey := range vs.schema {
+			if vs.schema[sKey] == storetypes.StoreTypePersistent {
+				sKeys = append(sKeys, sKey)
+			}
+		}
+
+		sort.Slice(sKeys, func(i, j int) bool {
+			return strings.Compare(sKeys[i], sKeys[j]) == -1
+		})
+
+		var storeByteKeys [][]byte
+		for _, sKey := range sKeys {
+			storeByteKeys = append(storeByteKeys, []byte(sKey))
+		}
+
+		err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+			Item: &storetypes.SnapshotItem_Schema{
+				Schema: &storetypes.SnapshotSchema{
+					Keys: storeByteKeys,
+				},
+			},
+		})
+		if err != nil {
+			chunkWriter.CloseWithError(err)
+			return
+		}
+
+		for _, sKey := range sKeys {
+			subStore, err := vs.getSubstore(sKey)
+			if err != nil {
+				chunkWriter.CloseWithError(err)
+				return
+			}
+
+			err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+				Item: &storetypes.SnapshotItem_Store{
+					Store: &storetypes.SnapshotStoreItem{
+						Name: sKey,
+					},
+				},
+			})
+			if err != nil {
+				chunkWriter.CloseWithError(err)
+				return
+			}
+
+			iter := subStore.Iterator(nil, nil)
+			for ; iter.Valid(); iter.Next() {
+				err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+					Item: &storetypes.SnapshotItem_KV{
+						KV: &storetypes.SnapshotKVItem{
+							Key:   iter.Key(),
+							Value: iter.Value(),
+						},
+					},
+				})
+				if err != nil {
+					chunkWriter.CloseWithError(err)
+					return
+				}
+			}
+			iter.Close()
+		}
+	}()
+
+	return ch, nil
```

Comment:
most of this code is copied. Let's do do that, and try to reuse ... one idea would be to move most of this checks to a common function to `/snapshots` package?

Reply:
ok

Reply:
Do you mean the

```go
if err != nil {
    chunkWriter.CloseWithError(err)
    return
}
```
?

There's 3 copies and they're only 1 line longer than a normal `err` check so that wouldn't be worth splitting into a separate package.

I can't find any other duplication here, or is it with some other files outside this diff that I am missing?

Reply:
Do you mean the

```go
if err != nil {
    chunkWriter.CloseWithError(err)
    return
}
```
?

There's 3 copies but they're only 1 line longer than a normal `err` check so that alone wouldn't be worth splitting into a separate package.

I can't find any other duplication here, or is it with some other files outside this diff that I am missing?